File: spa/src/three/streamClient/inputTrace.ts
// ============================================================================
//  three/streamClient/inputTrace — SAMPLED per-input tracing: the browser's
//  half of the in-record hop `docs/lab/TRACE-CONTEXT.md` documents and
//  `streamhost/src/input_trace.rs` decodes.
//  ---------------------------------------------------------------------------
//  THE DECISION IS MADE HERE, ONCE, and nowhere downstream samples again
//  (`scripts/serve/tracing.py`'s "SAMPLING IS THE BROWSER'S DECISION",
//  contract §5's W3C sampled-flag semantics applied to a transport that has no
//  flag byte of its own — presence of the suffix on the wire IS the flag). A
//  UI tab decides, mints a fresh trace, and appends its ids to the ~1-in-N
//  records this module is asked about; the daemon only ever RECOGNISES that
//  decision (`input_trace.rs`'s doc comment says so explicitly) and never
//  samples independently.
//
//  THE KNOB. `SAMPLE_N` below — one exported constant, one place, changed and
//  redeployed like any other SPA constant. Default 10: a visitor typing or
//  clicking normally produces a real span roughly once a second, enough to see
//  the input->pixel path without spending a span per keystroke.
//
//  WHAT NEVER TRAVELS. The suffix is 25 bytes: a marker and two ids. No
//  keycode, no button, no coordinate, no character is added to the record by
//  this module — those already exist in the record for the feature to work at
//  all (the guest has to be told which key), and this module never reads them
//  into a span attribute. `keyClass` below is a BUCKET the caller computes
//  from the same scancode already on its way to the wire, matching
//  `input_trace.rs::key_class` bucket-for-bucket so a browser span and a
//  daemon span drawn from the same edge agree on the word, without either
//  process ever transmitting which key it was.
// ============================================================================
import { startTrace, type Span } from '../../analytics/trace';
import { reportBackendTrace } from '../../analytics/instana';
import { transportAttrs } from './transportFacts';

// -- the EUM↔backend join (Instana `reportEvent`) ---------------------------
// A vendor beacon can name a backend trace; a backend span can never name a
// browser session — the join is one-directional and has to be driven from
// here, the ONE place a sampled `input.edge` trace is minted
// (docs/lab/TRACE-CONTEXT.md §3.2, docs/ANALYTICS.md §8.1).
//
// The mechanics — the guarded `ineum` call, the 16-or-32-hex `backendTraceId`
// rule the vendor enforces SILENTLY, the meta-key cap — now live in
// `analytics/instana.ts`, which is where facts about the vendor belong. They
// were declared here while this module was the only caller; `khFetch.ts`
// reporting the same join off a response header made a local copy a second
// opinion about what the vendor accepts, and two opinions is exactly how a
// silently-dropped field survives a green test suite.

/**
 * Tag a just-minted sampled `input.edge` trace onto the browser session via
 * Instana's EUM↔backend join. A no-op, entirely, when `window.ineum` does not
 * exist (unconfigured build — the vendor script never loaded) and when the
 * span is a NOOP (tracing off, empty trace id) — our own tracing above this
 * call is already complete and unaffected either way.
 *
 * `meta` NEVER carries a key's identity or typed text — the caller already
 * enforces that (`keyClass` is a bucket, `kh.input.class` is a wire-record
 * type), and this function only forwards what it is given.
 */
function reportSampledEdge(span: Span, meta: Record<string, string>): void {
  reportBackendTrace('kh.input.sampled', span.traceId, meta);
}

/** The knob: 1 input edge in this many gets a real span and a wire context.
 *  Documented default per docs/lab/TRACE-CONTEXT.md and docs/ANALYTICS.md §8.1
 *  — changing it is a one-line SPA edit, redeployed like any other constant. */
export const SAMPLE_N = 10;

/** 0xC5 marker + 16-byte trace id + 8-byte span id, matching
 *  `input_trace::SUFFIX_MARKER` / `SUFFIX_LEN` exactly. */
const SUFFIX_MARKER = 0xc5;
const TRACE_ID_BYTES = 16;
const SPAN_ID_BYTES = 8;
export const SUFFIX_LEN = 1 + TRACE_ID_BYTES + SPAN_ID_BYTES;

let counter = 0;

/** Reset the sampling counter. Test seam only — production never needs this,
 *  the counter free-runs for the life of the tab. */
export function __resetSampleCounter(): void {
  counter = 0;
  pendingEdges.clear();
  pendingOrder.length = 0;
}

/** The browser's sampling decision for ONE qualifying edge (a key transition
 *  or a button transition — never a pointer-move sample, which arrives at up
 *  to ~250 Hz and would defeat the entire point of sampling). Returns a live
 *  span on the sampled edge, `null` on the other N-1 — which is the ENTIRE
 *  cost an unsampled edge pays here: one increment, one comparison, no
 *  allocation, no id minted (`startTrace` is not even called). */
export function maybeSampleEdge(
  name: string,
  attrs: Record<string, string>,
): Span | null {
  counter += 1;
  if (counter < SAMPLE_N) return null;
  counter = 0;
  const span = startTrace(name, attrs, 'client');
  // Remember this edge so `frameTrace.ts` can close a browser-clock round trip
  // against the frame the daemon eventually names as its answer.
  if (span.traceId) {
    if (!pendingEdges.has(span.traceId)) pendingOrder.push(span.traceId);
    pendingEdges.set(span.traceId, { spanId: span.spanId, atMs: performance.now() });
    while (pendingOrder.length > MAX_PENDING_EDGES) {
      const old = pendingOrder.shift();
      if (old !== undefined) pendingEdges.delete(old);
    }
  }
  // Fires only on the ~1-in-SAMPLE_N edges that actually mint a trace, so the
  // other N-1 pay nothing here either — `reportSampledEdge` itself no-ops
  // below that (unconfigured build, or a NOOP span because our own tracing
  // is disabled).
  reportSampledEdge(span, attrs);
  return span;
}

// -- the transport hop, and the browser-clock round trip ---------------------
// TWO THINGS THE OLD SHAPE LEFT INVISIBLE, added here because this is the one
// module that already knows a sampled edge exists.
//
// 1. `input.wire` — the WebTransport hop as ITS OWN span rather than an
//    unexplained gap between `input.edge` (this tab) and `input.dispatch` (the
//    daemon). Its DURATION is only the local handoff: how long this tab spent
//    getting the record into the QUIC stream writer, which is where
//    backpressure shows up. The hop's actual cost is not a duration this tab
//    can measure — the daemon's clock is a different clock, and subtracting
//    two wall clocks across two machines yields skew, not latency — so it is
//    carried as the connection's own `kh.transport.rtt_ms` instead
//    (`transportFacts.ts`). A measured RTT beside a measured local enqueue is
//    an honest account of the hop; a span whose ends were read off two clocks
//    would not be.
//
// 2. `client.input.roundtrip` — edge to painted pixel, start and end BOTH read
//    from this tab's own `performance.now()`, so it is the one number in the
//    whole tree that needs no clock agreement at all. Emitted from
//    `frameTrace.ts` once the daemon names which `frame_id` answered this
//    edge. It is the operator's actual question ("how long until I saw it")
//    and it is the envelope the daemon's own spans decompose.

/** Edges awaiting their answering frame: `traceId` -> the edge span's id and
 *  the browser clock reading at which it was minted. Bounded, and an entry
 *  that never gets an answer simply ages out — the rest of the trace still
 *  stands. Small because the input side already samples 1-in-`SAMPLE_N` and a
 *  frame answers within a handful of frames. */
const MAX_PENDING_EDGES = 32;
const pendingEdges = new Map<string, { spanId: string; atMs: number }>();
const pendingOrder: string[] = [];

/** Consume the pending edge for `traceId`, if this tab minted one. Consuming
 *  (rather than peeking) is deliberate: exactly one round-trip span per edge,
 *  even if the daemon marks two frames against the same trace. */
export function takePendingEdge(traceId: string): { spanId: string; atMs: number } | null {
  const e = pendingEdges.get(traceId);
  if (!e) return null;
  pendingEdges.delete(traceId);
  return e;
}

/** Run `write` inside a child `input.wire` span when this edge is sampled, and
 *  run it untouched when it is not. `reliability` is the caller's own fact —
 *  `stream` for a key/button record on its per-class QUIC stream, `datagram`
 *  for pointer motion — because the two have genuinely different loss and
 *  latency behaviour.
 *
 *  The write is in a `try/finally` so a throwing writer still closes the span:
 *  a wire span that never ends would be a hole in the tree exactly where the
 *  failure was. */
export function writeTraced(
  span: Span | null,
  reliability: 'stream' | 'datagram',
  write: () => void,
): void {
  if (!span) { write(); return; }
  const wire = span.child('input.wire', transportAttrs(reliability), 'client');
  try {
    write();
  } finally {
    wire.end('ok');
  }
}

function hexToBytes(hex: string, out: Uint8Array, offset: number, len: number): void {
  for (let i = 0; i < len; i += 1) {
    out[offset + i] = parseInt(hex.substr(i * 2, 2), 16) || 0;
  }
}

/** Encode `span`'s ids as the 25-byte wire suffix. Only called for a span
 *  `maybeSampleEdge` actually returned, so this never runs on the unsampled
 *  path. */
export function traceSuffix(span: Span): Uint8Array {
  const out = new Uint8Array(SUFFIX_LEN);
  out[0] = SUFFIX_MARKER;
  hexToBytes(span.traceId, out, 1, TRACE_ID_BYTES);
  hexToBytes(span.spanId, out, 1 + TRACE_ID_BYTES, SPAN_ID_BYTES);
  return out;
}

/** Append `suffix` after `body`'s meaningful bytes (`body` may be a larger
 *  buffer than `bodyLen` when a caller over-allocated; only the first
 *  `bodyLen` bytes are kept). */
export function withSuffix(body: Uint8Array, bodyLen: number, suffix: Uint8Array): Uint8Array {
  const out = new Uint8Array(bodyLen + suffix.length);
  out.set(body.subarray(0, bodyLen), 0);
  out.set(suffix, bodyLen);
  return out;
}

/** `kh.key.class` — mirrors `input_trace::key_class` in streamhost bucket for
 *  bucket, computed from the SAME XT set-1 wire scancode
 *  (`streamClient/keysym.ts`) that is already being sent so the guest's
 *  keyboard can be driven at all. A bucket, never the key: two different keys
 *  in the same bucket produce the identical string. */
export function keyClass(scancode: number): string {
  const base = scancode & 0x00ff;
  const extended = (scancode >> 8) === 0xe0;
  if (base === 0x1c) return 'enter'; // Return, and extended 0xE01C numpad Enter
  const modifierBases = extended
    ? [0x1d, 0x38, 0x5b, 0x5c, 0x5d] // RCtrl/RAlt/LWin/RWin/Menu
    : [0x1d, 0x2a, 0x36, 0x38, 0x3a, 0x45, 0x46]; // Ctrl/LShift/RShift/Alt/Caps/Num/Scroll
  if (modifierBases.includes(base)) return 'modifier';
  const navBases = [0x0f, 0x47, 0x48, 0x49, 0x4b, 0x4d, 0x4f, 0x50, 0x51, 0x52, 0x53];
  if (navBases.includes(base)) return 'navigation';
  if ((!extended && base >= 0x3b && base <= 0x44) || base === 0x57 || base === 0x58) {
    return 'function';
  }
  return 'printable';
}

File: streamhost/streamhost/src/input_trace.rs
//! Sampled per-input trace context — the wire suffix, and the key-class
//! vocabulary it lets a span name without naming a key.
//!
//! THE DECISION IS THE BROWSER'S (`docs/lab/TRACE-CONTEXT.md` §5,
//! `scripts/serve/tracing.py`'s "SAMPLING IS THE BROWSER'S DECISION"). This
//! daemon never samples independently: it only recognises a context the
//! browser already minted and attached, and it never fabricates one when the
//! suffix is absent. The input plane has no headers (it is raw WebTransport
//! records, not HTTP — contract §3), so the only place left to carry a
//! `traceparent`-equivalent is inside the record itself, on the ~1-in-N
//! records the browser chose to sample.
//!
//! WIRE FORMAT. Appended AFTER a record's normal fixed fields, present only on
//! a sampled edge:
//!
//! ```text
//!   [ ...normal record bytes... ][ 0xC5 | trace-id (16 BE) | span-id (8 BE) ]
//!                                  ^^^^ SUFFIX_MARKER          25 bytes total
//! ```
//!
//! `input.rs`'s match arms all guard on `rec.len() >= N`, not `== N` — a
//! record with unrecognised trailing bytes is already handled as "ignore
//! them", which is exactly backward compatibility for free:
//!   * an OLD browser -> a NEW daemon sends records at their historic exact
//!     length; `strip` finds no suffix (the length doesn't match ANY known
//!     `base + SUFFIX_LEN`) and returns the record untouched with `None` —
//!     byte-identical to today.
//!   * a NEW browser -> an OLD daemon sends the longer record; the old
//!     `input::handle` never heard of this module, reads its fixed fields off
//!     the FRONT exactly as before and ignores the tail — the click or
//!     keystroke still lands, just untraced.
//!
//! Both directions are exercised in `input_trace_tests.rs`.
//!
//! WHY A MARKER BYTE AND NOT JUST A LENGTH CHECK. Two record shapes exist for
//! type 2 (the bare 3-byte button edge and the carried-position 11-byte one —
//! `input.rs`'s `sendButtonImpl` comment explains why), so `base + SUFFIX_LEN`
//! is not always unique against stray padding. The marker turns "the length
//! matches" into "the length matches AND the byte where a suffix would start
//! is the one only this code writes there" — a coincidence has to clear both.
//! A record that clears the length check but not the marker is treated as NO
//! suffix (never a hard error — contract §1's "malformed starts a new trace,
//! never refuses the work", applied to the one wire format this contract does
//! not already cover).
//!
//! COST WHEN NOT SAMPLED (9 records in 10 at the default N=10): `strip` is a
//! match on `rec[0]`, one or two length comparisons and (only when the length
//! already matches) one byte compare. No allocation, no formatting — the same
//! budget `probes.rs` and `trace_session.rs`'s `once()` hold themselves to.
//! `input_trace_suffix_strip_cost` in the test module measures it.

use crate::trace::Ctx;

/// Leads the 24-byte context payload when present. Arbitrary; chosen only to
/// be unlikely to collide with a coincidental trailing byte pattern.
pub const SUFFIX_MARKER: u8 = 0xC5;
/// `trace-id (16) + span-id (8)`, matching the 32-hex/16-hex halves of a
/// `traceparent` — no flags byte: presence of the suffix at all already means
/// "sampled", so encoding a redundant `01` would only be one more thing that
/// could disagree with itself.
const CTX_LEN: usize = 16 + 8;
/// The whole appended tail: marker + context.
pub const SUFFIX_LEN: usize = 1 + CTX_LEN;

/// Base (pre-suffix) lengths this daemon ever sends/receives for a wire
/// record TYPE, matching `input.rs`'s header table exactly. Only key (3) and
/// button (2) records ever carry a suffix — pointer motion (types 1/4) is
/// unreliable-datagram, out of scope by design (~250 samples/sec would blow
/// the whole point of sampling), and wheel/touch are not wired for it either.
fn base_lens(rec_type: u8) -> &'static [usize] {
    match rec_type {
        3 => &[4],     // key: u8 type, u8 down, u16 qemu_keycode
        2 => &[3, 11], // button: bare edge, or edge + carried position + cseq
        _ => &[],
    }
}

/// Split a possible trace suffix off the tail of `rec`. Returns the record
/// with the suffix removed (identical to `rec` when none was found) and the
/// context it carried, if any and if it parsed to a non-zero trace/span id
/// (an all-zero id is the same "invalid" spelling `trace/context.rs` already
/// rejects for the header form).
pub fn strip(rec: &[u8]) -> (&[u8], Option<Ctx>) {
    let Some(&rec_type) = rec.first() else {
        return (rec, None);
    };
    for &base in base_lens(rec_type) {
        if rec.len() != base + SUFFIX_LEN {
            continue;
        }
        if rec[base] != SUFFIX_MARKER {
            continue;
        }
        let trace_bytes: [u8; 16] = rec[base + 1..base + 17].try_into().unwrap();
        let span_bytes: [u8; 8] = rec[base + 17..base + 25].try_into().unwrap();
        let trace = u128::from_be_bytes(trace_bytes);
        let span = u64::from_be_bytes(span_bytes);
        if trace == 0 || span == 0 {
            return (&rec[..base], None);
        }
        return (
            &rec[..base],
            Some(Ctx {
                trace,
                span,
                sampled: true,
            }),
        );
    }
    (rec, None)
}

/// Encode a suffix for `ctx` — the daemon never sends one today (only the
/// browser mints sampled contexts), but the test module round-trips through
/// this to prove `strip` is its exact inverse rather than merely "looks
/// right".
#[cfg(test)]
pub fn encode(ctx: Ctx) -> [u8; SUFFIX_LEN] {
    let mut out = [0u8; SUFFIX_LEN];
    out[0] = SUFFIX_MARKER;
    out[1..17].copy_from_slice(&ctx.trace.to_be_bytes());
    out[17..25].copy_from_slice(&ctx.span.to_be_bytes());
    out
}

/// `kh.input.class` — the coarse WIRE class, never the payload.
pub fn input_class(rec_type: u8) -> &'static str {
    match rec_type {
        2 => "click",
        3 => "key",
        _ => "other",
    }
}

/// `kh.key.class` — a small, deliberately coarse vocabulary computed from the
/// XT set-1 wire scancode (`input.rs`'s type=3 payload — the SAME code
/// `key()` already injects into the guest, already necessary for the feature
/// to work at all). This is a BUCKET, not the key: two different keys sharing
/// a bucket produce the identical string, and nothing here can be inverted
/// back to which key was pressed. Extended keys ride as `0xE0xx` on the wire
/// (`key_quirks.rs`'s `key_qnum` doc comment); the bucket checks the base
/// byte, which is why the bare-keypad and enhanced-extended forms of the same
/// physical key land in the same class.
pub fn key_class(qemu_keycode: u16) -> &'static str {
    let base = (qemu_keycode & 0x00FF) as u8;
    let extended = (qemu_keycode >> 8) == 0xE0;
    if base == 0x1C {
        return "enter"; // main Return, and extended 0xE01C numpad Enter
    }
    let modifier = if extended {
        matches!(base, 0x1D | 0x38 | 0x5B | 0x5C | 0x5D) // RCtrl/RAlt/LWin/RWin/Menu
    } else {
        matches!(base, 0x1D | 0x2A | 0x36 | 0x38 | 0x3A | 0x45 | 0x46)
    };
    if modifier {
        return "modifier";
    }
    // Tab plus the bare-keypad/enhanced navigation cluster — both encodings
    // share these base bytes (see the module doc), so extended is irrelevant.
    if matches!(
        base,
        0x0F | 0x47 | 0x48 | 0x49 | 0x4B | 0x4D | 0x4F | 0x50 | 0x51 | 0x52 | 0x53
    ) {
        return "navigation";
    }
    if !extended && (0x3B..=0x44).contains(&base) || base == 0x57 || base == 0x58 {
        return "function";
    }
    "printable"
}

#[cfg(test)]
#[path = "input_trace_tests.rs"]
mod tests;
